### PR TITLE
Add `Upgrade-Insecure-Requests` header model

### PR DESCRIPTION
**Pull request by diogocanut**
Adds `Upgrade-Insecure-Requests` as part of #5010 .

Let me know what can be improved.

### CODE REVIEW

File: core/shared/src/main/scala/org/http4s/headers/`Upgrade-Insecure-Requests`.scala
Comment (satorg) [inline core/shared/src/main/scala/org/http4s/headers/`Upgrade-Insecure-Requests`.scala:1]: Would you remove backticks from the filename please? All other files in `scala/org/http4s/headers` do not use backticks (which simplifies navigation among them quite a bit).

File: core/shared/src/main/scala/org/http4s/headers/`Upgrade-Insecure-Requests`.scala [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import cats.parse.Parser
+import cats.parse.Parser.char
+import org.typelevel.ci._
+
+sealed abstract class `Upgrade-Insecure-Requests`(val value: String)
+    extends Product
+    with Serializable
+
+object `Upgrade-Insecure-Requests` {
+  case object Upgrade extends `Upgrade-Insecure-Requests`("1")
```
Comment (satorg) [inline core/shared/src/main/scala/org/http4s/headers/`Upgrade-Insecure-Requests`.scala:29]: Looks like it is just that:
```
Upgrade-Insecure-Requests: 1
```
i.e. it always takes "1" and no other choice is acceptable. In other words, it is either the header is provided or not.

Therefore, perhaps we could keep it even simpler, something like that:
```scala
sealed trait `Upgrade-Insecure-Requests`

case object `Upgrade-Insecure-Requests` extends `Upgrade-Insecure-Requests` {
  ...
}
```
I.e. no inner object would be necessary, wdyt?

Comment (diogocanut) [reply core/shared/src/main/scala/org/http4s/headers/`Upgrade-Insecure-Requests`.scala:29]: I agree, it always 1. I'll change it.

Comment (diogocanut) [reply core/shared/src/main/scala/org/http4s/headers/`Upgrade-Insecure-Requests`.scala:29]: Should we still have the parser for `1`?

```
  private[http4s] val parser: Parser[`Upgrade-Insecure-Requests`] =
    char('1').as(`Upgrade-Insecure-Requests`)
```

**Review by satorg — APPROVED** [review]
Thank you for the PR!

File: core/shared/src/main/scala/org/http4s/headers/Upgrade-Insecure-Requests.scala [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package org.http4s
+package headers
+
+import cats.parse.Parser
+import cats.parse.Parser.char
+import org.typelevel.ci._
+
+sealed trait `Upgrade-Insecure-Requests`
```
Comment (armanbilge) [inline core/shared/src/main/scala/org/http4s/headers/Upgrade-Insecure-Requests.scala:24]:
```suggestion
sealed abstract class `Upgrade-Insecure-Requests`
```

File: tests/shared/src/test/scala/org/http4s/headers/UpgradeInsecureRequestsSuite.scala [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+import cats.syntax.either._
+import org.http4s.laws.discipline.arbitrary._
+
+class UpgradeInsecureRequestsSuite extends HeaderLaws {
+  checkAll("Upgrade-Insecure-Requests", headerLaws[`Upgrade-Insecure-Requests`])
+
+  test("parsing 1 into Upgrade-Insecure-Requests") {
+    assertEquals(
+      `Upgrade-Insecure-Requests`.parser.parseAll("1"),
```
Comment (armanbilge) [inline tests/shared/src/test/scala/org/http4s/headers/UpgradeInsecureRequestsSuite.scala:28]:
```suggestion
      `Upgrade-Insecure-Requests`.parse("1"),
```

**Review by armanbilge — COMMENTED** [review]
Just some nitpicks :)

File: core/shared/src/main/scala/org/http4s/headers/Upgrade-Insecure-Requests.scala (hunk 2 above)
Comment (armanbilge) [reply core/shared/src/main/scala/org/http4s/headers/Upgrade-Insecure-Requests.scala]: Honestly it's my personal preference, with this reasoning:

1. I don't see a practical reason it needs to be a `trait` (this is true for most ADTs in my experience)
2. `abstract class`es have theoretically better performance than `trait`s. in any case, it's a cheaper encoding
3. in Scala 3 `enum` is encoded as an `abstract class`, presumably for similar reasons. it also means we could theoretically rewrite as `enum` without breaking binary-compatibility

**Review by armanbilge — APPROVED** [review]
Nice one, thank you!

**Pull request merged**